src/main/scala/com/cognite/sdk/scala/common/OAuth2.scala

```diff
@@ -21,19 +27,37 @@ object OAuth2 {
       audience: Option[String] = None
   )
 
+  final case class Session(
+      sessionKey: String,
+      cdfProjectName: String,
+      tokenFromVault: String
+  )
+
+  private def commonGetAuth[F[_]](cache: CachedResource[F, TokenState])(
+      implicit F: Monad[F],
+      clock: Clock[F]
+  ): F[Auth] = for {
+    now <- clock.monotonic(TimeUnit.SECONDS)
+    _ <- cache.invalidateIfNeeded(_.expiresAt <= now)
+    auth <- cache.run(state => F.pure(OidcTokenAuth(state.token, state.cdfProjectName)))
+  } yield auth
+
   class ClientCredentialsProvider[F[_]] private (
       cache: CachedResource[F, TokenState]
   )(
       implicit F: Monad[F],
       clock: Clock[F]
   ) extends AuthProvider[F]
       with Serializable {
-    def getAuth: F[Auth] =
-      for {
-        now <- clock.monotonic(TimeUnit.SECONDS)
-        _ <- cache.invalidateIfNeeded(_.expiresAt <= now)
-        auth <- cache.run(state => F.pure(OidcTokenAuth(state.token, state.cdfProjectName)))
-      } yield auth
+    def getAuth: F[Auth] = commonGetAuth(cache)
+  }
+
+  class SessionProvider[F[_]] private (
+      cache: CachedResource[F, TokenState]
+  )(implicit F: Monad[F], clock: Clock[F])
+      extends AuthProvider[F]
+      with Serializable {
+    def getAuth: F[Auth] = commonGetAuth(cache)
```

Comment:
Wondering if these can be smashed together even if there's a companion object for each.. but can't think of how!
They're formatted differently for some reason anyways :raised_eyebrow:

Reply:
yes I prefer to separate those for now to avoid breaking stuff 😅 , we can merge them later anyway 😁